framework/routing/__init__.py

```diff
@@ -187,15 +198,20 @@ def render_jinja_string(tpl, data):
     pass
 
 mako_cache = {}
-def render_mako_string(tpldir, tplname, data):
+def render_mako_string(tpldir, tplname, data, safe=False):
+    """Render a mako template. Optional safe argument to activate markupsafe escaping"""
+    lookup_obj = _TPL_LOOKUP_SAFE if safe is True else _TPL_LOOKUP
 
     tpl = mako_cache.get(tplname)
     if tpl is None:
+        with open(os.path.join(tpldir, tplname)) as f:
```

Comment:
Oooh - good catch.

This sort of thing I believe is the root of the "Too many open files" errors many of us have been getting while running the OSF.